### ISSUE TITLE
Running coverage on a contract which uses a contract without ContractDefinition fails

### ISSUE THREAD

**Issue by robin-thomas** [sc-forks/solidity-coverage#429]
If you have a smart contract A.sol, whose coverage is skipped in `.solcover.js`, as shown below (which is used in the truffle tests), then the unit tests are not getting run when we run the coverage command (but the tests work fine with `truffle test` command).

```
pragma solidity >=0.5.0 <0.6.0;

import "npm_package/contracts/A.sol";
```

But if I copy the file `npm_package/contracts/A.sol` into the contract file A.sol and run the coverage, then it works.

Any idea how to solve this (without copying over the solidity contract from node_modules)?

I'm using version `0.7.0-beta.2`.

**Comment by robin-thomas** [comment]
Hi @cgewecke,

* Correct. The smart contract file has only an import statement. The use case for this is only for testing. I'll explain more.

Let's say, my main contract needs to mock the operations of an external contract (`A.sol`) only during testing. During real deployment, my smart contract will be talking to the external contract that's maintained by that company (so I already have the contract address for that external contract), which runs behind their network. So for testing my smart contract, I want to mock all that. So I take their `A.sol` file and then deploy that along with my smart contract, and then mocks its functionalities.

Check out https://github.com/robin-thomas/flyt/src/test/truffle/flyt.js to get some idea.

* The smart contracts are in `src/truffle/contracts`. If I replace the file `src/truffle/contracts/lib/Oracle.sol` to the what's shown below,
```
pragma solidity >=0.5.0 <0.6.0;

import "chainlink/v0.5/contracts/Oracle.sol
```
the error I reported happens.

**Comment by cgewecke** [comment]
@robin-thomas Ah great, thanks for reporting this. Wasn't familiar with that usage.... 

I will add that case to our tests and make sure it's handled. The beta will probably not be published again for ~10 days, so until then you will have to use a work-around. Apologies.

**Comment by robin-thomas** [comment]
@cgewecke Ah no issues. I can use my workaround for now.

Thanks for this great project anyway!

**Comment by cgewecke** [comment]
Just leaving some notes about this because I think it will remain open for a bit....

There was an undefined property crash in SC because the possibility of a contract file without a root node was not considered. That is fixed in #429. 

*However*, the test I created for this fails anyway :/ There's a quirk in the truffle compile method for this case which I can't figure out. Error looks like this.
```
Error: the string "TypeError: Cannot convert undefined or null to object\n    at Function.entries (<anonymous>)\n    at processContracts (/Users/cgewecke/code/sc-forks/v2/solidity-coverage/node_modules/truffle/build/webpack:/packages/truffle-compile/run.js?cc41:335:1)\n    at run (/Users/cgewecke/code/sc-forks/v2/solidity-coverage/node_modules/truffle/build/webpack:/packages/truffle-compile/run.js?cc41:65:1)" was thrown, throw an Error :)
```

A work-around seems to be to add a 'no-op' contract definition to the pure import file. For example
```solidity
pragma solidity >=0.4.21 <0.6.0;

import "npmpackage/ChainLink.sol";

interface Void {}
```

`ChainLink.json` is created successfully with that and can be `artifacts.require`'d without any problems.

**Comment by cgewecke** [comment]
It looks like this was a bug in Truffle that's been fixed (not sure where). 

Have added a working test case for this in #454, after upgrading Truffle dev dependency to 5.1.2. 

Closing, please feel free to ping if you continue to see this.
